upup/pkg/fi/cloudup/awstasks/natgateway.go

```diff
@@ -0,0 +1,151 @@
+limitations under the License.
+*/
+
+package awstasks
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/golang/glog"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/awsup"
+	"k8s.io/kops/upup/pkg/fi/cloudup/terraform"
+)
+
+//go:generate fitask -type=natgateway
+type NATGateway struct {
+	ID           *string // AWS ID for the object
+	AllocationID *string // Pointer to an elastic IP
+	SubnetID     *string // Pointer to a subnet in the VPC
+	VPCID        *string // Unique identifier of the VPC for this NAT gateway
+}
+
+var _ fi.CompareWithID = &NATGateway{} // Validate the IDs
+
+func (e *NATGateway) CompareWithID() *string {
+	return e.ID
+}
+
+func (e *NATGateway) Find(c *fi.Context) (*NATGateway, error) {
+	cloud := c.Cloud.(awsup.AWSCloud)
+
+	request := &ec2.DescribeNatGatewaysInput{}
+
+	if fi.StringValue(e.ID) != "" {
+		request.NatGatewayIds = []*string{e.ID}
+	} else {
+		request.Filter = cloud.BuildFilters(e.SubnetID)
+	}
+
+	response, err := cloud.EC2().DescribeNatGateways(request)
+	if err != nil {
+		return nil, fmt.Errorf("error listing NAT Gateways: %v", err)
+	}
+	if response == nil || len(response.NatGateways) == 0 {
+		return nil, nil
+	}
+
+	if len(response.NatGateways) != 1 {
+		return nil, fmt.Errorf("found multiple NAT Gateways matching tags")
+	}
+	ngw := response.NatGateways[0]
+	actual := &NATGateway{
+		ID:   ngw.NatGatewayId,
+		VPCID: ngw.VpcId,
+		SubnetID: ngw.SubnetId,
+	}
+
+	glog.V(4).Infof("found matching NAT gateway %v", actual)
+
+	// Allocation ID
+	if actual.ID != nil {
+		request := &ec2.DescribeAddressesInput{}
+		request.Filters = cloud.BuildFilters(e.VPCID)
+		response, err := cloud.EC2().DescribeAddresses(request)
+		if err != nil || len(response.Addresses) != 1 {
+			return nil, fmt.Errorf("error querying for elastic ip support: %v", err)
+		}
+		actual.AllocationID = response.Addresses[0].AllocationId
+	}
+
+	if e.ID == nil {
+		e.ID = actual.ID
+	}
+
+	return actual, nil
+}
+
+func (s *NATGateway) CheckChanges(a, e, changes *NATGateway) error {
+	if a == nil {
+		if e.AllocationID == nil {
+			return fi.RequiredField("AllocationID")
+		}
+		if e.SubnetID == nil {
+			return fi.RequiredField("SubnetID")
+		}
+	}
+	if a != nil {
+		if changes.AllocationID != nil {
+			// TODO: Do we want to destroy & recreate the VPC?
+			return fi.CannotChangeField("AllocationID")
+		}
+	}
+	return nil
+}
+
+func (e *NATGateway) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(e, c)
+}
+
+func (_ *NATGateway) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *NATGateway) error {
+	if a == nil {
+		glog.V(2).Infof("Creating NGW with Allocation ID: %q", *e.AllocationID)
+
+		request := &ec2.CreateNatGatewayInput{
+			AllocationId: e.AllocationID,
+			SubnetId: e.SubnetID,
+		}
+
+		response, err := t.Cloud.EC2().CreateNatGateway(request)
+		if err != nil {
+			return fmt.Errorf("error creating Nat gateway: %v", err)
```

Comment:
What I've tried doing (inconsistently) is referring to them by the class name i.e. `NATGateway` everywhere.